### PR TITLE
Jeb Bearer and Daniel Zhang

### PR DESCRIPTION
**Pull request by hmc-cs-jbearer**

### CODE REVIEW

File: language-design.md [hunk 1]
```diff
@@ -17,7 +17,82 @@ For each quote, describe what it was about the quote that led you pick it.
...
+
+We also disagree with the sentiment that programmers should employ syntax more
+intuitive to novices.  Sure, using natural language could look more
+understandable for someone who has never coded before.  However, for more
+experienced coders, using natural language could feel ambiguous and cumbersome.
+Frequent users of code would appreciate a concise set of keywords that they
+could build off of.  Also, we think it is more important to cater to experienced
+coders than the "n00bs" in the study because they were only novices for an hour.
+Typically, people spend more time as experienced coders who understand the
+syntax than as beginners trying to read off code like a natural language.  There
```
Comment (hmc-cs-rparamastri) [inline language-design.md:43]: Your comment on "trying to read off code like a natural language" made me think about how I read code. A lot of people sound out each word in their head when they read natural language. I do a similar thing in math, with some symbol heavy exceptions such as when reading {a_1, ..., a_n}, which I read as "a set containing a one through a n."

For code, I don't read off every symbol when I see a line such as

```
for (int i = 0; i < 10; ++i) {
}
```

but I do read it in "natural language". I think I would read this loop as "a for loop that runs 10 times." But I may be overthinking this concept, I'm not sure if I actually sound it out like that in my head. But I am sure that I don't read every symbol, because when confronted with a non-idiomatic loop such as

```
for (int i = 0; i <= 10; ++i) {
}
```

I slow down and pay more attention to what the for loop is actually stating. I'm curious to know what other people "hear" in their head when they read a line of code.

Comment (MichaelSheely) [reply language-design.md:43]: Renata makes a great point here! This description is so similar to my own thought process when I read code I just had to jump in and give my experience as well. My brain definitely automatically pattern matches idiomatic for loops into a natural language "do 10 times", while a non-idiomatic one takes a little time to ensure I understand what it is saying.  I suppose one fear here is that my eyes might miss such a subtle difference from an idiomatic for loop, potentially leading me to misread code if I don't see the equal sign.

File: language-design.md [hunk 2]
```diff
@@ -17,7 +17,82 @@ For each quote, describe what it was about the quote that led you pick it.
...
+They then utilize the language prejudice to keep the outsiders out and
+themselves in. Breaking this cycle of privilege will require a dramatic
+cultural shift.
+
+3. From the abstract by Bloch:
+
+> "You can't please everyone, so aim to displease everyone equally" [Bloch,
+2006]
+
+We chose this quote because it is an undesirable, yet important part of
```
Comment (hmc-cs-rparamastri) [inline language-design.md:80]: This is an interpretation of the quote that I hadn't thought of, but I totally agree with this. I thought the quote related more to Steele's idea on creating small languages — you can't include every feature that different people want, so you instead make a language that is growable.

File: language-design.md [hunk 3]
```diff
@@ -28,7 +103,54 @@ you know a poorly designed language? What are the symptoms?
 
 **Response**
 
-
+Guy Steele gives us several traits of a good language in "Growing a Language".
+According to Steele, the size of the language is one important indicator. As he
```
Comment (hmc-cs-rparamastri) [inline language-design.md:107]: As a follow up to this, how would you determine the size of a language. Like what is too small and what's too large?
I agree with your answer that a small and growable language is preferred, but what are some identifiers of languages that are just the right size? You mention some features that a growable language would have, but I think you should also mention how one would judge the size of a language.

File: language-design.md [hunk 4]
```diff
@@ -38,18 +160,89 @@ How do the themes of _Growing a Language_ relate to the "sound lab" we did this
...
+to make their own functions that look like primitives of the language. 
+According to Steele, 
+> APL was designed by one man, a smart man—and I love APL—but it had a flaw that
+> I think has all but killed it: there was no way for a user to grow the
+> language in a smooth way (Steele, 6)
+
+In retrospect, we failed to consider this point when designing our Sound 
+API.  Our design involved creating a Sound class with two data members, the 
+sound samples and the sampling rate.  Our goal was to keep these data members 
+private and only expose users to the class's methods.  (Bloch includes this
```
Comment (hmc-cs-rparamastri) [inline language-design.md:198]: I don't think Python supports private data members. I did some research and I found [this post](https://plus.google.com/115212051037621986145/posts/7wpbQTPRWft) by Guido van Rossum (Python's "Benevolent Dictator for Life"), which explains some of his reasoning behind not supporting private data members. In fact, Python seems to encourage directly accessing data members. Getters and setters are not considered "pythonic".

I think Bloch's maxim that data members should be private is a principle shared by many people in CS, so it's interesting how the Python community prefers otherwise. I'm biased towards Python, so I'm not sure if this is a count against Python being a good language.

Comment (hmc-cs-jbearer) [reply language-design.md:198]: I think it's fairly common to use an underscore before a variable name to indicate that it's private. Obviously users could still access it if they really want to, but by convention it means something along the lines of "this is an implementation detail; you probably shouldn't mess with it".

The reason we wanted our data to be treated as private is so that if we later decide to represent samples in a different way (maybe by working in the frequency domain instead of the time domain) it shouldn't break any of the code that uses this API, since samples being represented as lists + sampling rate is not part of the interface. Then the cost of maintaining that implementation-interface barrier is that users can't easily subclass our Sound class, which means they cannot extend the API with new member functions, which ultimately means user-added functions will look slightly different than built-in ones.

Interesting response by Guido. I don't know if I agree with him, but then again I'm C++ biased. It just seems to me that the additional safety of having the language (compiler or interpreter) alert the programmer when they're accessing stuff they shouldn't be is a lot to give up for making the debugger slightly more powerful.

File: language-design.md (hunk 3 above)
Comment (hmc-cs-jbearer) [reply language-design.md:107]: Good point. Honestly, I'm not really sure how to answer this. I'm tempted to say that with enough experience a programmer should be able to tell when their language is too small or too large, even if they can't articulate why. But that's not a useful answer. A possible indicator of a language being too large might be that new users have a hard time learning all of it, but there are so many confounding factors when it comes to speed of language-learning. For too-small languages, maybe we could look for something Steele mentioned: does the user have to implement many functions that feel like they should be primitives? That's already kind of nebulous, though. We wouldn't say ContextFree is too small because someone trying to do data analysis has to do a lot of overhead first, because ContextFree isn't concerned with data analysis. Even for a general purpose language, it's hard to tell when something should be added. The many types of numbers mentioned by Steele may feel like they should be primitives to someone working with numbers, but that doesn't mean they all belong in the language.

tl;dr I'm not really sure

File: language-design.md [hunk 5]
```diff
@@ -59,7 +252,35 @@ What does the post on grayscale tell us about the process of API design?
...
+Least Astonishment: "Every method should do the least surprising thing it
+could, given its name" [Bloch, 2006].  The writers of the API went to great
+lengths to provide the least astonishing name for their function by polling
+users to gauge their intuition.  They also considered whether small changes in
+behavior would require extensive changes in the code. This provided
+encouragement for `rgb(x)`, since if you later wanted to change the color away
+from gray, you would only have to change the arguments rather than the function
+name.  Furthermore, readability was a big concern for naming the grayscaling
+function.  Although `rgb(x)` may have its benefits, it is much less readable
+than, for example, `gray(x)`.
```
Comment (hmc-cs-rparamastri) [inline language-design.md:276]: I agree with this. Verou argues that since programmers are accustomed to using `rgb()`, it would be beneficial to use the same function name for grayscale. However, if I saw a function call to `rgb()` that only uses one argument, I wouldn't immediately understand that it's for grayscale.

File: language-design.md [hunk 6]
```diff
@@ -102,6 +365,72 @@ you do so? If not, why not?
...
+This block resembles code that we are familiar with, but some of the braces and
+parentheses have been replaced with words.  On the other hand, here is a
+snippet of AppleScript (from The Ultimate Beginner's Guide To AppleScript,
+Josh Johnson, 2012):
+
+```
+[applescript]
+tell application "Mail"
+
+set theMessage to make new outgoing message with properties
```
Comment (hmc-cs-rparamastri) [inline language-design.md:395]: I really like how you included an example of AppleScript. This line also shows the extent to which natural language is incorporated in the language. I can definitely see how writing AppleScript could be confusing.

File: language-design.md [hunk 7]
```diff
@@ -17,7 +17,82 @@ For each quote, describe what it was about the quote that led you pick it.
...
+have trouble intuiting code then it's not because you're a n00b.  However,
+they go on to reference a studying that tested people who have never coded
+before.  This quote does not support their point that experienced coders
+struggle with syntax.
+
+We also disagree with the sentiment that programmers should employ syntax more
+intuitive to novices.  Sure, using natural language could look more
+understandable for someone who has never coded before.  However, for more
+experienced coders, using natural language could feel ambiguous and cumbersome.
+Frequent users of code would appreciate a concise set of keywords that they
```
Comment (hmc-cs-dobermiller) [inline language-design.md:39]: What are you basing this statement off of? Personal experience/preference? A study? While I don't disagree, a little justification (or simply saying "likely") would be nice.

File: language-design.md [hunk 8]
```diff
@@ -17,7 +17,82 @@ For each quote, describe what it was about the quote that led you pick it.
...
+could build off of.  Also, we think it is more important to cater to experienced
+coders than the "n00bs" in the study because they were only novices for an hour.
+Typically, people spend more time as experienced coders who understand the
+syntax than as beginners trying to read off code like a natural language.  There
+is certainly a niche for languages that resemble natural communication, for
+example a stepping stone to teach beginners computer science or a domain-
+specific language not intended for people with prior CS experience.  However,
+once a student has learned how to code, he or she will grow tired of the
```
Comment (hmc-cs-dobermiller) [inline language-design.md:47]: This still seems to make a lot of assumptions, and I'm not even sure that it would be true. A lot of BDD and TDD now use very verbose, natural languagey test frameworks because they're easier to explain tests in (see Gherkin, Cucumber, etc).

File: language-design.md [hunk 9]
```diff
@@ -17,7 +17,82 @@ For each quote, describe what it was about the quote that led you pick it.
...
 **Response**
 
-
+1. From the article by Pavlus:
+
+>"According to findings by researchers from Southern Illinois University, this
+>reaction isn’t just because you’re a n00b: they found that Perl, a major
+>programming language used by untold zillions of developers, is no more
+>intuitive to novices than a language with a randomly generated syntax."
+>[Pavlus, 2012].
```
Comment (hmc-cs-dobermiller) [inline language-design.md:26]: Some google searching (credit to @danielsonner @mbae for finding this) reveals that "randomly generated syntax" is a bit of a stretch

> A programming language based largely on the syntactical structure of Quorum. 
> With the exception of braces, the lexical rule for variable names, and a few 
> operators (e.g., addition, subtraction, multiplication, division), many of the
> keywords and symbols were chosen randomly from the ASCII table. 
>  [[Stefik, 20000]](https://ecs.victoria.ac.nz/foswiki/pub/Events/PLATEAU/Program/plateau2011-stefik.pdf)

File: language-design.md [hunk 10]
```diff
@@ -17,7 +17,82 @@ For each quote, describe what it was about the quote that led you pick it.
...
+Both of us spent the summer working at tech companies whose engineers were
+overwhelmingly male, so this quote struck us as being particularly accurate.
+Sadly, it epitomizes the current state of the tech industry as a place
+bifurcated into groups of "insiders" and "outsiders", one in which ability and
+potential have little to do with which group one finds themselves in.
+
+A particularly interesting point of the article, which we believe is
+encapsulated in this quote, is the ambiguous direction of causality when it
+comes to language-based discrimination. Do so-called "Real Programmers" look
+down on Javascript or HTML users because of prejudice against those languages?
```
Comment (hmc-cs-dobermiller) [inline language-design.md:66]: A lot of it also has to do with accessibility - you'll find tons of articles and write-ups about how making programming more accessible is ruining programming by letting in a bunch of unskilled n00bs, and I think that web development is one of the more accessible ones. And they're right to an extent - the more novice programmers there are, the more bad code will be written. But I think the point they miss is that, eventually if people stick with it, they'll stop writing bad code.

There are arguments against it - [Jeff Atwood has gone on many rants](https://blog.codinghorror.com/separating-programming-sheep-from-non-programming-goats/) about n00bs in software development (that's not the post I was looking for, but it's a decent start). Regardless, I think that there is a lot of embedded elitisim in software development.

File: language-design.md (hunk 2 above)
Comment (hmc-cs-dobermiller) [inline language-design.md:80]: I disagree with the quote, to an extent. Trying to displease everyone equally seems a little silly, especially when there are a whole bunch of people you weren't trying to please anyway. In particular for a DSL, you should be aiming at displeasing domain experts equally, and ignoring the effect on everyone else. Maybe I'm being a bit too pedantic with my interpretation, but there are still some people who you should displease less than others, and that group is your target audience.

File: language-design.md [hunk 11]
```diff
@@ -38,18 +160,89 @@ How do the themes of _Growing a Language_ relate to the "sound lab" we did this
...
 **Response**
 
-
+Our main takeaway from _Growing a Language_ was that a good language should
+provide users with a small and useful set of building blocks.  From there, the
+programmer should be able to compose the available tools to suit their needs. We
+tried to use this mentality when adapting the Sound library to a more usable
+API.  First, we created a "Sound" class which contained all of the methods for
+editing sounds.  We tried to be consistent--each method took in a Sound (self)
+along with other parameters and returned a new Sound.  This way, the user could
```
Comment (hmc-cs-dobermiller) [inline language-design.md:169]: Was there a reason you chose immutability? I.e. why not stack all of the changes in a single `Sound` instance?

File: language-design.md [hunk 12]
```diff
@@ -38,18 +160,89 @@ How do the themes of _Growing a Language_ relate to the "sound lab" we did this
...
+
+In addition, _Growing a Language_ showed the importance of keeping an API as 
+contained as possible, and only including options that would be relevant to 
+many users.  On pages 10 and 11, Steele shows many examples of data types, each
+increasingly obscure.  After describing them, Steele asks "So should we make
+'x' a type in the Java programming language?"  In the end, he concludes that
+language designers should not add these rare features because "It would not
+be fair to weigh down all programmers with the need to have or to learn all the
+words for all niche uses" (Steele, 12).  This idea ties with the theme of making
+a language/API exactly as small as it should be, and no smaller [Bloch, 2006].
```
Comment (hmc-cs-dobermiller) [inline language-design.md:180]: I think a language like Python does this pretty well - the builtin functionality of the language is relatively limited, but a massive stdlib is provided alongside it.

File: language-design.md (hunk 4 above)
Comment (hmc-cs-dobermiller) [reply language-design.md:198]: The `_` prefix also prevents discovery by various tools for introspection (unless you try a bit harder). Additionally, getters and setters are unnecessary in Python because they use properties instead, i.e.

``` python
class Foo:
    _bar = None

    @property
    def bar(self):
        if self._bar is None:
            self._bar = self.expensive_initialization_call()
        return self._bar

    @bar.setter
    def bar(self, value):
        if self.validate(value):
            self._bar = value
        else:
            raise ValueError("{} failed validation.".format(value))  

a = Foo()
print a.bar
a.bar = 42
```

It isn't that you can't or shouldn't use encapsulation in Python - it's just that there isn't a huge benefit to preventing users from accessing private values. If they choose to do so that's fine, but it's at their own risk.

Additionally, properties make Python much easier to extend than Java. With Java, the main reason we have `getX` and `setX` methods is because if we started at a `public int` but eventually wanted to calculate it instead we have a breaking interface change, thus we use `private int` and getters and setters so we don't have those breaking changes. With Python there are no breaking interface changes even if we want to make a property lazily evaluated, or provide extra validation logic, or whatever.

File: language-design.md (hunk 5 above)
Comment (hmc-cs-dobermiller) [reply language-design.md:276]: At the same time, for domain experts it should be immediately understandable (once they learn about it, at least). For the target audience, it makes sense.

File: language-design.md [hunk 13]
```diff
@@ -59,7 +252,35 @@ What does the post on grayscale tell us about the process of API design?
...
+from gray, you would only have to change the arguments rather than the function
+name.  Furthermore, readability was a big concern for naming the grayscaling
+function.  Although `rgb(x)` may have its benefits, it is much less readable
+than, for example, `gray(x)`.
+
+Lastly, this article reveals, as Bloch writes, "You can't please everyone so 
+aim to displease everyone equally" [Bloch, 2006].  The comments show that many
+users had strong opinions about how to name the grayscaling function. 
+Therefore, many programmers will be displeased no matter what name was chosen. 
+It was impossible to satisfy everybody in this instance, so the best the
```
Comment (hmc-cs-dobermiller) [inline language-design.md:282]: I commented on this earlier as well, but it occurs to me that there are occasionally times when as the designers and developers you have to make your own decision, even if you end up displeasing more people; sometimes you really do know best, and sometimes you need to make unpopular choices.

File: language-design.md [hunk 14]
```diff
@@ -70,7 +291,49 @@ what ways do the themes apply to the study and creation of DSLs?
 
 **Response**
 
-
+There are clearly stereotypes about certain DSLs. For example, Yang and Rabkin
+mention the tendency of back-end developers to look down on front-end
+programmers. They mention Javascript in this context. While Javascript is
+clearly a general purpose language, there are DSLs involved in web development
+that also contribute to the stereotypes. For example, HTML fits our definition
+of a DSL, and more generally a programming language. It has syntax and
```
Comment (hmc-cs-dobermiller) [inline language-design.md:299]: Calling HTML a programming language is a huge stretch.  The vast number of articles you mention below make a lot of compelling arguments; I think you'll need to give some more justification on this point.

File: language-design.md [hunk 15]
```diff
@@ -70,7 +291,49 @@ what ways do the themes apply to the study and creation of DSLs?
...
+> language or markup language is a bit blurry.
+
+> Apparently, HTML5+CSS3 is turing complete too.
+
+
+What's interesting here is not whether these languages are Turing-complete, or
+if they are, well, languages. Rather, the interesting thing is that Turing-
+completeness is an issue of discussion at all. There seems to be a popular
+conception that Turing-completeness makes a programming language, and since we
+have seen that many DSLs are not Turing-complete, it seems reasonable to
```
Comment (hmc-cs-dobermiller) [inline language-design.md:319]: Great point!
